quickwit-indexing/src/actors/garbage_collector.rs

```diff
@@ -0,0 +1,278 @@
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+use std::time::Duration;
+
+use async_trait::async_trait;
+use quickwit_actors::{Actor, ActorContext, AsyncActor};
+use quickwit_metastore::Metastore;
+use quickwit_storage::Storage;
+use tracing::info;
+
+use crate::run_garbage_collect;
+
+const RUN_INTERVAL: Duration = Duration::from_secs(10 * 60); // 10 minutes
+const GRACE_PERIOD: Duration = Duration::from_secs(2 * 60); // 2 minutes
+
+#[derive(Debug, Clone, Default)]
+pub struct GarbageCollectorCounters {
+    /// The number of passes the garbage collector has performed.
+    pub num_passes: usize,
+    /// The number of deleted files.
+    pub num_deleted_files: usize,
+    /// The number of bytes deleted.
+    pub num_deleted_bytes: usize,
+    /// The number of failed to delete files.
+    pub num_failed_files: usize,
+}
+
+/// An actor for collecting garbage periodically from an index.
+pub struct GarbageCollector {
+    index_id: String,
+    index_storage: Arc<dyn Storage>,
+    metastore: Arc<dyn Metastore>,
+    counters: GarbageCollectorCounters,
+}
+
+impl GarbageCollector {
+    pub fn new(
+        index_id: String,
+        index_storage: Arc<dyn Storage>,
+        metastore: Arc<dyn Metastore>,
+    ) -> Self {
+        Self {
+            index_id,
+            index_storage,
+            metastore,
+            counters: GarbageCollectorCounters::default(),
+        }
+    }
+}
+
+impl Actor for GarbageCollector {
+    type Message = ();
+    type ObservableState = GarbageCollectorCounters;
+
+    fn observable_state(&self) -> Self::ObservableState {
+        self.counters.clone()
+    }
+}
+
+#[async_trait]
+impl AsyncActor for GarbageCollector {
+    async fn initialize(
+        &mut self,
+        ctx: &ActorContext<Self::Message>,
+    ) -> Result<(), quickwit_actors::ActorExitStatus> {
+        self.process_message((), ctx).await
+    }
+
+    async fn process_message(
+        &mut self,
+        _: (),
+        ctx: &ActorContext<Self::Message>,
+    ) -> Result<(), quickwit_actors::ActorExitStatus> {
+        info!(index = %self.index_id, "garbage-collect-operation");
+        self.counters.num_passes += 1;
+
+        let deletion_stats = run_garbage_collect(
```

Comment:
is `run_garbage_collect` during more than a heartbeat? If yes, I'm worried about this actor becoming unhealthy.

Reply:
I don't get this question?

Reply:
(discussed offline)